README.md
<p align="center">
  <img src="https://user-images.githubusercontent.com/11348/56806359-b6a07f00-67f9-11e9-96bd-6a456a96880c.png" width="600" />
  <br /><br />
  <code>PlugCanonicalHost</code> ensures that all requests are served by a single canonical host.<br /> It will redirect all requests from non-canonical hosts to the canonical one.
  <br /><br />
  <a href="https://github.com/remi/plug_canonical_host/actions?query=workflow%3ACI+branch%3Amaster+event%3Apush"><img src="https://github.com/remi/plug_canonical_host/workflows/CI/badge.svg?branch=master&event=push" /></a>
  <a href="https://hex.pm/packages/plug_canonical_host"><img src="https://img.shields.io/hexpm/v/plug_canonical_host.svg" /></a>
</p>

## Installation

Add `plug_canonical_host` to the `deps` function in your project’s `mix.exs` file:

```elixir
defp deps do
  [
    …,
    {:plug_canonical_host, "~> 2.0"}
  ]
end
```

Then run `mix do deps.get, deps.compile` inside your project’s directory.

## Usage

`PlugCanonicalHost` can be used just as any other plugs. Add `PlugCanonicalHost` before all of the other plugs you want to happen after successful redirect to your canonical host.

The recommended way to define a canonical host is with an environment variable.

```elixir
# config/releases.exs
config :my_app,
  canonical_host: System.get_env("CANONICAL_HOST")

# lib/my_app/endpoint.ex
defmodule MyApp.Endpoint do
  plug(:canonical_host)

  defp canonical_host(conn, _opts) do
    :my_app
    |> Application.get_env(:canonical_host)
    |> case do
      host when is_binary(host) ->
        opts = PlugCanonicalHost.init(canonical_host: host)
        PlugCanonicalHost.call(conn, opts)

      _ ->
        conn
    end
  end
end
```

For example, if your `CANONICAL_HOST` is `www.example.com` but your application is accessible via both `example.com` and `www.example.com`, all traffic coming through `example.com` will be redirected (with a `301` HTTP status) to the matching `www.example.com` URL.

```bash
$ curl -sI "http://example.com/foo?bar=1"
#> HTTP/1.1 301 Moved Permanently
#> Location: http://www.example.com/foo?bar=1
```

### Allow multiple hosts

If you want to support _multiple_ hosts, you can make them skip the canonical host behavior but you’re still going to have to specify a canonical host for unknown hosts redirects.

```elixir
# config/releases.exs
config :my_app,
  canonical_host: System.get_env("CANONICAL_HOST"),
  passthrough_hosts: String.split(System.get_env("PASSTHROUGH_HOSTS"), ",")

# lib/my_app/endpoint.ex
defmodule MyApp.Endpoint do
  plug(:canonical_host)

  defp canonical_host(%Plug.Conn{host: host} = conn, _opts) do
    canonical_host = Application.get_env(:app, :canonical_host)
    passthrough_hosts = Application.get_env(:my_app, :passthrough_hosts)

    cond do
      host in passthrough_hosts ->
        conn
      is_binary(canonical_host) ->
        opts = PlugCanonicalHost.init(canonical_host: canonical_host)
        PlugCanonicalHost.call(conn, opts)
      true ->
        conn
    end
  end
end
```

Let’s say we have `CANONICAL_HOST=www.example.com` and `PASSTHROUGH_HOSTS=foo.example.com,bar.example.com`.

Now, all requests going through `www.example.com`, `foo.example.com` or `bar.example.com` path will skip the canonical host redirect behavior. Other hosts will redirect to `www.example.com`.

```bash
$ curl -sI "http://example.com/foo?bar=1"
#> HTTP/1.1 301 Moved Permanently
#> Location: http://www.example.com/foo?bar=1

$ curl -sI "http://foo.example.com/foo?bar=1"
#> HTTP/1.1 200 OK

$ curl -sI "http://bar.example.com/foo?bar=1"
#> HTTP/1.1 200 OK

$ curl -sI "http://www.example.com/foo?bar=1"
#> HTTP/1.1 200 OK
```

### Exclude certain requests

If you want to _exclude_ certain requests from redirecting to the canonical host, you can use simple pattern matching in your function arguments:

```elixir
# config/releases.exs
config :my_app,
  canonical_host: System.get_env("CANONICAL_HOST")

# lib/my_app/endpoint.ex
defmodule MyApp.Endpoint do
  plug(:canonical_host)

  defp canonical_host(%Plug.Conn{request_path: "/ignore-me"} = conn, _opts) do
    Plug.Conn.send_resp(conn, 200, "👋")
  end

  defp canonical_host(conn, _opts) do
    :my_app
    |> Application.get_env(:canonical_host)
    |> case do
      host when is_binary(host) ->
        opts = PlugCanonicalHost.init(canonical_host: host)
        PlugCanonicalHost.call(conn, opts)

      _ ->
        conn
    end
  end
end
```

Now, all requests going to the `/ignore-me` path will skip the canonical host redirect behavior.

```bash
$ curl -sI "http://example.com/foo?bar=1"
#> HTTP/1.1 301 Moved Permanently
#> Location: http://www.example.com/foo?bar=1

$ curl -sI "http://example.com/ignore-me"
#> HTTP/1.1 200 OK
```

## License

`PlugCanonicalHost` is © 2016-2020 [Rémi Prévost](http://exomel.com) and may be freely distributed under the [MIT license](https://github.com/remi/plug_canonical_host/blob/master/LICENSE.md). See the `LICENSE.md` file for more information.

The plug logo is based on [this lovely icon by Vectors Market](https://thenounproject.com/term/usb-plug/298582), from The Noun Project. Used under a [Creative Commons BY 3.0](http://creativecommons.org/licenses/by/3.0/) license.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

Since it is a boilerplate project, there are technically no official (versioned) _releases_. Therefore, the `master` branch should always be stable and usable.

## v2.0.3

- Downgrade `earmark` to make CI build pass with older Elixir versions

## v2.0.2

- Upgrade `ex_doc` and `earmark` to fix raw HTML in documentation

## v2.0.1

- Fix various warnings for Elixir 1.11 (#19, thanks @gcauchon!)

## v2.0.0

- Remove `ignore` since the same behaviour is achievable with a simple Plug setup.

## Previous versions

See `git log` for a complete history. Sorry about that.


test/plug_canonical_host_test.exs
defmodule PlugCanonicalHostTest do
  use ExUnit.Case, async: true
  use Plug.Test

  defmodule TestApp do
    use Plug.Router

    plug(PlugCanonicalHost, canonical_host: "example.com")
    plug(:match)
    plug(:dispatch)

    get "/foo" do
      conn |> send_resp(200, "Hello World")
    end
  end

  test "redirects to canonical host" do
    conn =
      :get
      |> conn("http://www.example.com/")
      |> TestApp.call(TestApp.init([]))

    assert conn.status == 301
    assert get_resp_header(conn, "location") === ["http://example.com/"]
  end

  test "redirects to canonical host with query string" do
    conn =
      :get
      |> conn("http://www.example.com/foo?bar=1")
      |> TestApp.call(TestApp.init([]))

    assert conn.status == 301
    assert get_resp_header(conn, "location") === ["http://example.com/foo?bar=1"]
  end

  test "redirects to forwarded port" do
    conn =
      :get
      |> conn("https://www.example.com/foo?bar=1")
      |> Map.put(:port, 80)
      |> put_req_header("x-forwarded-port", "443")
      |> TestApp.call(TestApp.init([]))

    assert conn.status == 301
    assert get_resp_header(conn, "location") === ["https://example.com/foo?bar=1"]
  end

  test "redirects to forwarded proto" do
    conn =
      :get
      |> conn("http://www.example.com/foo?bar=1")
      |> put_req_header("x-forwarded-proto", "https")
      |> put_req_header("x-forwarded-port", "443")
      |> TestApp.call(TestApp.init([]))

    assert conn.status == 301
    assert get_resp_header(conn, "location") === ["https://example.com/foo?bar=1"]
  end

  test "redirects to forwarded proto without forwarded port" do
    conn =
      :get
      |> conn("http://www.example.com/foo?bar=1")
      |> put_req_header("x-forwarded-proto", "https")
      |> TestApp.call(TestApp.init([]))

    assert conn.status == 301
    assert get_resp_header(conn, "location") === ["https://example.com/foo?bar=1"]
  end

  test "does not redirect to canonical host when already on canonical host" do
    conn =
      :get
      |> conn("http://example.com/foo")
      |> TestApp.call(TestApp.init([]))

    assert conn.status == 200
    assert conn.resp_body == "Hello World"
  end

  test "does not redirect when canonical host is an empty string" do
    conn =
      %Plug.Conn{host: "www.example.com", status: 200}
      |> PlugCanonicalHost.call(canonical_host: "")

    assert conn.status == 200
  end

  test "does not redirect when canonical host is nil" do
    conn =
      %Plug.Conn{host: "www.example.com", status: 200}
      |> PlugCanonicalHost.call(canonical_host: nil)

    assert conn.status == 200
  end
end


mix.exs
defmodule PlugCanonicalHost.Mixfile do
  use Mix.Project

  @version "2.0.3"

  def project do
    [
      app: :plug_canonical_host,
      version: @version,
      elixir: "~> 1.8",
      deps: deps(),
      package: package(),
      dialyzer: dialyzer(),
      name: "Plug Canonical Host",
      source_url: "https://github.com/remi/plug_canonical_host",
      homepage_url: "https://github.com/remi/plug_canonical_host",
      description:
        "PlugCanonicalHost ensures that all requests are served by a single canonical host. It will redirect all requests from non-canonical hosts to the canonical one.",
      docs: [extras: ["README.md"], main: "readme", source_ref: "v#{@version}", source_url: "https://github.com/remi/plug_canonical_host"]
    ]
  end

  def application do
    [
      applications: [:plug]
    ]
  end

  def dialyzer do
    [plt_add_apps: [:plug], plt_file: ".plts/.local.plt", plt_core_path: ".plts"]
  end

  defp deps do
    [
      {:plug, " ~> 1.0"},
      {:credo, "~> 0.10", only: :dev},
      {:earmark_parser, "1.4.12", only: :dev},
      {:ex_doc, "~> 0.24", only: :dev},
      {:dialyxir, "~> 0.5", only: :dev}
    ]
  end

  defp package do
    %{
      maintainers: ["Rémi Prévost"],
      licenses: ["MIT"],
      links: %{
        "GitHub" => "https://github.com/remi/plug_canonical_host"
      }
    }
  end
end


.credo.exs
%{
  configs: [
    %{
      name: "default",
      strict: true,
      files: %{
        included: ["lib/", "test/", "priv/"],
        excluded: []
      },
      checks: [
        {Credo.Check.Consistency.ExceptionNames},
        {Credo.Check.Consistency.LineEndings},
        {Credo.Check.Consistency.SpaceAroundOperators},
        {Credo.Check.Consistency.SpaceInParentheses},
        {Credo.Check.Consistency.TabsOrSpaces},

        {Credo.Check.Design.AliasUsage},
        {Credo.Check.Design.DuplicatedCode, excluded_macros: []},
        {Credo.Check.Design.TagTODO, false},
        {Credo.Check.Design.TagFIXME, false},

        {Credo.Check.Readability.FunctionNames},
        {Credo.Check.Readability.LargeNumbers},
        {Credo.Check.Readability.MaxLineLength, max_length: 200},
        {Credo.Check.Readability.ModuleAttributeNames},
        {Credo.Check.Readability.ModuleDoc, false},
        {Credo.Check.Readability.ModuleNames},
        {Credo.Check.Readability.ParenthesesInCondition},
        {Credo.Check.Readability.PredicateFunctionNames},
        {Credo.Check.Readability.TrailingBlankLine},
        {Credo.Check.Readability.TrailingWhiteSpace},
        {Credo.Check.Readability.VariableNames},

        {Credo.Check.Refactor.ABCSize},
        {Credo.Check.Refactor.CaseTrivialMatches},
        {Credo.Check.Refactor.CondStatements},
        {Credo.Check.Refactor.FunctionArity},
        {Credo.Check.Refactor.MatchInCondition},
        {Credo.Check.Refactor.PipeChainStart,
          excluded_argument_types: ~w(atom binary fn keyword)a,
          excluded_functions: ~w(from)},
        {Credo.Check.Refactor.CyclomaticComplexity},
        {Credo.Check.Refactor.NegatedConditionsInUnless},
        {Credo.Check.Refactor.NegatedConditionsWithElse},
        {Credo.Check.Refactor.Nesting},
        {Credo.Check.Refactor.UnlessWithElse},

        {Credo.Check.Warning.IExPry},
        {Credo.Check.Warning.IoInspect, false},
        {Credo.Check.Warning.OperationOnSameValues},
        {Credo.Check.Warning.BoolOperationOnSameValues},
        {Credo.Check.Warning.UnusedEnumOperation},
        {Credo.Check.Warning.UnusedKeywordOperation},
        {Credo.Check.Warning.UnusedListOperation},
        {Credo.Check.Warning.UnusedStringOperation},
        {Credo.Check.Warning.UnusedTupleOperation},
        {Credo.Check.Warning.OperationWithConstantResult},
      ]
    }
  ]
}


lib/plug_canonical_host.ex
defmodule PlugCanonicalHost do
  @moduledoc """
  A Plug for ensuring that all requests are served by a single canonical host
  """

  # Imports
  import Plug.Conn

  # Aliases
  alias Plug.Conn

  # Behaviours
  @behaviour Plug

  # Constants
  @location_header "location"
  @forwarded_port_header "x-forwarded-port"
  @forwarded_proto_header "x-forwarded-proto"
  @status_code 301
  @html_template """
    <!DOCTYPE html>
    <html lang="en-US">
      <head><title>301 Moved Permanently</title></head>
      <body>
        <h1>Moved Permanently</h1>
        <p>The document has moved <a href="%s">here</a>.</p>
      </body>
    </html>
  """

  # Types
  @type opts :: binary | tuple | atom | integer | float | [opts] | %{opts => opts}

  @doc """
  Initialize this plug with a canonical host option.
  """
  @spec init(opts) :: opts
  def init(opts) do
    [
      canonical_host: Keyword.fetch!(opts, :canonical_host)
    ]
  end

  @doc """
  Call the plug.
  """
  @spec call(%Conn{}, opts) :: Conn.t()
  def call(conn = %Conn{host: host}, canonical_host: canonical_host)
      when is_nil(canonical_host) == false and canonical_host !== "" and host !== canonical_host do
    location = conn |> redirect_location(canonical_host)

    conn
    |> put_resp_header(@location_header, location)
    |> send_resp(@status_code, String.replace(@html_template, "%s", location))
    |> halt
  end

  def call(conn, _), do: conn

  @spec redirect_location(%Conn{}, String.t()) :: String.t()
  defp redirect_location(conn, canonical_host) do
    conn
    |> request_uri
    |> URI.parse()
    |> sanitize_empty_query()
    |> Map.put(:host, canonical_host)
    |> URI.to_string()
  end

  @spec request_uri(%Conn{}) :: String.t()
  defp request_uri(conn = %Conn{host: host, request_path: request_path, query_string: query_string}) do
    "#{canonical_scheme(conn)}://#{host}:#{canonical_port(conn)}#{request_path}?#{query_string}"
  end

  @spec canonical_port(%Conn{}) :: binary | integer
  defp canonical_port(conn = %Conn{port: port}) do
    case {get_req_header(conn, @forwarded_port_header), get_req_header(conn, @forwarded_proto_header)} do
      {[forwarded_port], _} -> forwarded_port
      {[], ["http"]} -> 80
      {[], ["https"]} -> 443
      {[], []} -> port
    end
  end

  @spec canonical_scheme(%Conn{}) :: binary
  defp canonical_scheme(conn = %Conn{scheme: scheme}) do
    case get_req_header(conn, @forwarded_proto_header) do
      [forwarded_proto] -> forwarded_proto
      [] -> scheme
    end
  end

  @spec sanitize_empty_query(%URI{}) :: %URI{}
  defp sanitize_empty_query(uri = %URI{query: ""}), do: Map.put(uri, :query, nil)
  defp sanitize_empty_query(uri), do: uri
end
